File: classes01/cpp/opening.cpp
#include <opencv2/opencv.hpp>

using namespace std;
using namespace cv;

int main(void)
{
    Mat image=imread("../../data/images/opening.png",IMREAD_GRAYSCALE);

    int iterations=3;

    int erosionSize=3;
    Mat element=getStructuringElement(MORPH_ELLIPSE,Size(2*erosionSize+1,2*erosionSize+1),Point(erosionSize,erosionSize));

    Mat imageMorphOpened;
    morphologyEx(image,imageMorphOpened,MORPH_OPEN,element,Point(-1,-1),iterations);

    Mat imageDilation;
    dilate(image,imageDilation,element,Point(-1,-1),iterations);

    Mat imageErosion;
    erode(image,imageErosion,element,Point(-1,-1),iterations);

    imshow("Original", image);
    imshow("Dilation",imageDilation);
    imshow("Erosion",imageErosion);
    imshow("Opening",imageMorphOpened);

    waitKey(0);
    destroyAllWindows();

    return EXIT_SUCCESS;
}

File: classes01/python/opening.py
import cv2

image=cv2.imread("../../data/images/opening.png",cv2.IMREAD_GRAYSCALE)

# 检查是否是有效的输入
if image is None:
    print("读取文件失败")

openingSize=3
element = cv2.getStructuringElement(cv2.MORPH_ELLIPSE,(2*openingSize+1,2*openingSize+1),(openingSize,openingSize))

imageMorphOpened=cv2.morphologyEx(image,cv2.MORPH_OPEN,element,iterations=3)

cv2.imshow("Original", image)
cv2.imshow("Opening", imageMorphOpened)

# 测试代码
imageMorphClosed=cv2.morphologyEx(image,cv2.MORPH_CLOSE,element,iterations=3)
cv2.imshow("Closed",imageMorphClosed)

cv2.waitKey(0)
cv2.destroyAllWindows()

